### ISSUE TITLE
Error reporting from plugins, and log.error

### ISSUE THREAD

**Issue by oprypin** [mkdocs/mkdocs#2082]
I've been wondering how to properly report warnings and errors that happen during operation of my plugin.

I found (not through documentation) that perhaps it should be this:

https://github.com/mkdocs/mkdocs/blob/44f3ae212d85a60750769df0e9fa8022274166c8/mkdocs/structure/files.py#L7-L11 https://github.com/mkdocs/mkdocs/blob/44f3ae212d85a60750769df0e9fa8022274166c8/mkdocs/structure/files.py#L241

(with the *very* subtle issue that mkdocs configures only `mkdocs.*` loggers but the plugin would not fall under that and would hence not follow the `--verbose` option and such.)

What's nice is that warnings automatically cause the generation to fail in MkDocs `strict` mode.

But I was surprised that `log.error` does *not* cause a failure **at all** even in `strict` mode (so I guess it looks for *warning* specifically), though I could argue that it should cause a failure even in non-strict mode. But I guess it's controversial to change the default behavior, and there's something to be said about errors causing an immediate crash versus being able to wait until the end and then cause a failure.

MkDocs itself seems to always employ a combination of `log.error` + `raise` (sometimes `raise SystemExit`). But I don't think a plugin can just `raise` anything like that, it'll cause a stack trace to show up, which MkDocs makes sure to avoid in its own handling, but a plugin can't prevent that, other than intentionally raising `SystemExit`, which is not great. It's *also* impossible to implement an error that should not be allowed to go unnoticed but also doesn't have to cause an immediate crash (config validation errors are the one example where that *does* work).

So to sum up, the use case of "log an error now and make sure to fail at the end" seems impossible to do from a plugin, and the use case of "log an error now and fail now without a stack trace" is kinda ugly.
And either way, error reporting for plugins is under-documented. All of the nuance above is certainly not obvious.

**Comment by pawamoy** [comment]
I agree. 

At first I was using `mkdocs.utils.log` because I thought it was designed for this. In fact, it's just the utils logger. Each MkDocs module has its own. So I changed my code to get the loggers myself, with `log = logging.getLogger(__name__)`, and now my loggers are not initialized like MkDocs' loggers., and so do not follow the verbose option. I don't feel like duplicating initialization code in all plugins would be a good idea. So I will go back to `from mkdocs.utils import log`.

I also wanted to respect the strict mode for warnings and errors, and found out about `warning_filters`, which is very nice and easy to use! But then realized I *had* to raise an exception for errors because they are not counted like warnings are. And it's problematic for a few reasons:
- MkDocs will catch the exception and log an error. I don't need that second generic error, I already logged one with a specific message explaining the error.
- MkDocs will then re-raise the exception, aborting the build here, and will not execute the `on_post_build` hook, which I need in any case for tearing things down.
- The traceback adds no value. It comes from something I handled, for which I logged an error, so I know why it happened. Example: I currently have to raise a `ValueError` when `site_url` is not set (after the error log of course) to make the build fail. This traceback is completely useless :confused:. Even worse, it could drive users to open an issue ("traceback? -> unusual error -> ticket").

I don't know MkDocs' code enough to offer a solution, but the ideas that easily come to mind are:
- add an `error_filter` to count errors? fail if > 1, just like warnings? MkDocs would still be free to raise exception internally.
- provide an official `log` object for plugins to import from mkdocs and use? so they don't have to import `warning_filter`, `error_filter` and add them to the logger?

**Comment by oprypin** [comment]
> So I will go back to `from mkdocs.utils import log`.

That's basically equivalent to `logging.getLogger('mkdocs.utils')`, and in that case I'd say `logging.getLogger('mkdocs.mkdocs_my_super_plugin')` is strictly better. Which is my plan for now.

The next part of your message just expands slightly on my message. Good point about cleanups.

And regarding the last part. The log objects *should* be created separately, to distinguish them per module. Maybe provide a function to create one by name, though. The current state is fine too, just confusing in terms of namespaces.

**Comment by pawamoy** [comment]
> logging.getLogger('mkdocs.mkdocs_my_super_plugin')

Cool, didn't know we could do that, thanks :slightly_smiling_face: What's the rule here, does the name just have to start with `mkdocs`?

**Comment by oprypin** [comment]
Yes, because `mkdocs` is being configured in main and that affects everything under it.
https://github.com/mkdocs/mkdocs/blob/3bada392d2430e31edd55d67b59d60899c22ae11/mkdocs/__main__.py#L25-L26

To clarify, these namespaces should represent the actual module name, and someone's plugin is not actually *part of* mkdocs namespace, so this is not a best practice still.

**Comment by waylan** [comment]
For a bit of context, the error system existed before any plugin system was designed, let alone implemented. And the only plugin which any core dev has worked on is the `search` plugin, which is included with MkDocs. In other words, we don't have a good point of reference for the needs of plugin devs here. If any changes can be made to the error system which helps plugin devs, we are open to a PR.

**Comment by pawamoy** [comment]
@waylan thanks for your answer! Then I'll try and send a PR, unless you would like to handle it @oprypin?

**Comment by oprypin** [comment]
I'd send a pull request for a function that constructs a logger, that's not controversial. The rest really needs to be discussed first. E.g. I'd like to make `log.error` cause an error at the end, and regardless of strict mode.

**Comment by oprypin** [comment]
> Currently, to get an error you `raise` one. This means that the error happens immediately 

Yes but then a stack trace output is unavoidable if coming from a plugin, while MkDocs has its private ways to prevent a stack trace.

Can we introduce an exception type that plugins would use and MkDocs would catch and silence?

**Comment by pawamoy** [comment]
I'd also need the `on_post_build` event to always be triggered, but maybe this belong in a new issue.

**Comment by waylan** [comment]
The CLI is only catching and hiding tracebacks for `mkdocs.exceptions.ConfigurationError` (and `OSError` for the server). Any other errors assume debugging will be needed and include a full traceback. Presumably the same behavior would be utilized by a plugin. What other types of errors are you guys raising in your plugins?

**Comment by oprypin** [comment]
Well, any error that is not configuration!
Say, there's a plugin to paste contents of another file, and that file doesn't exist.

**Comment by oprypin** [comment]
> Any other errors assume debugging will be needed and include a full traceback. Presumably the same behavior would be utilized by a plugin.

Ah, I had checked for some other kind of errors and saw that there's no traceback in this example: making a file intentionally unreadable `chmod -r foo.md` only prints `[Errno 13] Permission denied: <filename>` without traceback.
But it turns out that I ran into the very specific case that it happened to be `OSError` and within `mkdocs serve`!
https://github.com/mkdocs/mkdocs/blob/f532ade3ab1ad4beedbe3815e7771f60481f4836/mkdocs/__main__.py#L138-L140 Generally you're right though.

**Comment by waylan** [comment]
> Say, there's a plugin to paste contents of another file, and that file doesn't exist.

That sound like a configuration error to me. Or are you thinking that as the link to the "included" file is defined in the body of a Markdown document, that that is not configuration but something else?

These are the sorts of things we need feedback on. But a blanket state which essentially says, 'the current system is insufficient' isn't helpful. We need concrete examples. I have been dodging your suggests thought this discussion because you have only been speaking in generalities without specific concrete examples. I can only guess what problems you are really trying to resolve. Remember, from my perspective the system is complete. So provide well thought out real-world examples of scenarios which are not addressed. Then provide concrete proposals as solutions.

**Comment by pawamoy** [comment]
Alright, here's my concrete example :grin: 

My plugin renders Jinja templates. Users can override these templates with custom ones. If they try to `{% include` a template that does not exist, Jinja will raise a `jinja2.exceptions.TemplateNotFound`. while rendering them. In that case, I catch the exception and log an error. I don't let the exception bubble up because its stack trace is not useful (it does not contain the directory listing of the Jinja environment loader), instead I log a single error line. Like you said @waylan, there's no point in continuing to build: it should abort right now and fail. But for this, the only way is to raise an error, which I avoided in the first place.

My plugin also uses its own Markdown extension. The extension injects the HTML rendered by Jinja back into the XML tree with `xml.etree.ElementTree.XML`. If the HTML cannot be parsed correctly, XML raises a `ParseError`. If it happens, I catch the exception and log an error. Just like before, I don't let it bubble up, because the stack trace is not useful: it doesn't contain the text that XML failed to parse. In the error I log, I show the line and character that caused a parse error. It actually happened a few days ago: [markdown-katex was not properly closing a link tag, causing XML to fail](https://github.com/pawamoy/mkdocstrings/issues/87).

I admit these are cosmetic details: I want the build to fail and I don't want to print a traceback for it.
But it also happens that raising an exception will prevent MkDocs from running my `on_post_build` hook, which I need in any case to terminate opened background processes and invalidate caches. This is especially required when serving the site with livereload and watched directories.

To summarize, the actual behavior:

- `log.warning`: build continues ✓ `on_post_build` runs ✓ build fails ✓
- `log.error` without raise: build doesn't stop ✗ `on_post_build` runs ✓ build doesn't fail ✗
- `log.error` with raise: build stops ✓ `on_post_build` doesn't run ✗ build fails ✓ (cosmetic details: extra error line and noisy traceback ✗)

Of course, I'm not saying MkDocs should never catch and re-raise exceptions: it is extremely useful for *unexpected* errors.

What I need :pray: :

- [x] `log.warning`: build continues ✓ `on_post_build` runs ✓ build fails ✓
- [ ] (`log.error`: build continues ✓ `on_post_build` runs ✓ build fails ✓)<br>actual behavior + `error_filter`
- [ ] `log.error` + way to abort the build: build stops ✓ `on_post_build` runs ✓ build fails ✓

Finally, I realize I'm asking for several things that should maybe be discussed separately, though I think they are closely related. Also, @oprypin, @waylan, I hope I'm not hijacking this issue. If you feel so, I'll gladly move this comment to the one I opened a few days ago, or open a new issue on your advice @waylan :slightly_smiling_face:

**Comment by oprypin** [comment]
What you want is also what I want :)
Good writeup

**Comment by oprypin** [comment]
Hm sorry, I withdraw from making a pull request about this. I thought it'd be clear how to do it, but I don't know :)

**Comment by pawamoy** [comment]
I guess we could start with counting errors to make the build fail at the end, even if it's not aborted by an exception. Then we could add a new exception in mkdocs, that when caught would abort the build without printing the traceback, but still execute the `on_post_build` hook. Or if this change is too impactful, we could simply add a new `on_build_error`. I like it better actually, it's very explicit about its purpose and when its executed.

**Comment by pawamoy** [comment]
I have a working draft.

```console
% mkdocs build
INFO    -  Cleaning site directory 
INFO    -  Building documentation to directory: /media/data/dev/pawamoy/mkdocstrings/site 
ERROR   -  Error reading page 'reference/extension.md': mkdocstrings.extension: mismatched tag: line 456, column 5 (character f):
<p></f>
If your Markdown contains angle brackets < >, try to wrap them between backticks `< >`, or replace them with &lt; and &gt; 

Aborted build due to error.
```

With this I don't even have to log an error since MkDocs will log one with the `BuildError` message :+1:

**Comment by waylan** [comment]
> I catch the exception and log an error. I don't let the exception bubble up because its stack trace is not useful (it does not contain the directory listing of the Jinja environment loader), instead I log a single error line. 

By the way, the same concerns exists a few times in MkDocs code. Note how that is handled here:

https://github.com/mkdocs/mkdocs/blob/3bada392d2430e31edd55d67b59d60899c22ae11/mkdocs/structure/pages.py#L126-L135

In the case of either error above, the error message is not helpful. Therefore, we generate a helpful message, log it, and then re-raise the exception to stop MkDocs from running.

https://github.com/mkdocs/mkdocs/blob/399f8428b84c4a91a00e0388d7e9bb5260dc8ec3/mkdocs/commands/build.py#L180-L182

https://github.com/mkdocs/mkdocs/blob/399f8428b84c4a91a00e0388d7e9bb5260dc8ec3/mkdocs/commands/build.py#L228-L230

Those last two are interesting for plugin devs as they both catch `Exception`, which is Python's base error class. In other words they will catch any and all errors raised. And the `try` blocks in both cases include most (all?) of the calls to the "page" plugin events. Therefore, if a plugin raises an error of any kind from `on_pre_page`, `on_page_markdown`, `on_page_content`, `on_page_context`, `on_post_page`, we grab the error message from it and log that error. Then we re-raise the exception to stop MkDocs from running.

My suggestion would be for a plugin to do something similar. If the message in an exception raised by Jinja isn't helpful, catch it, generate a helpful error message and re-raise with the helpful message. MkDocs will then catch that exception and log the message along with the page it is associated with. 

I realize this doesn't help with non-page events. But perhaps the change would be to wrap all events in similar try/except blocks. Of without making any changes to MkDocs, you could log the message yourself and then re-raise. I realize this doesn't address counting errors, but in cases where the build should stop at the first error, it is already possible to get what you are asking for.

**Comment by waylan** [comment]
```python
except BuildError as e:
    # Run `build_error` plugin events.
    config['plugins'].run_event('build_error', error=e)
    raise SystemExit('\nAborted build due to error.')
```

First of all this completely hides the underlying error and makes debugging impossible. Catching an error, modifying the error message, and logging it are all reasonable goal, but not at the cost of debugging. That will only frustrate both core and plugin developers. I suppose we could include a catch-all which re-raised errors only in verbose mode. However, we haven't had many issued with users being overwhelemed by the tracebacks we are not hiding so I'm not sure it's necessary.

Second, what purpose is a `build_error` event supposed to serve? In other words, what would a plugin use it for? Sorry, but I don't follow here.

**Comment by waylan** [comment]
> Second, what purpose is a `build_error` event supposed to serve? In other words, what would a plugin use it for? Sorry, but I don't follow here.

Oh, is that to address this:

> But it also happens that raising an exception will prevent MkDocs from running my `on_post_build` hook, which I need in any case to terminate opened background processes and invalidate caches. This is especially required when serving the site with livereload and watched directories.

I would think opening background processes in a context manager (using `with`) would resolve that. But I can't think of any reason to invalidate caches. For that matter, it's not clear to me why you need to open background processes either. Clearly you are attempting something much more complex that anything I ever envisioned.

**Comment by pawamoy** [comment]
> My suggestion would be for a plugin to do something similar. If the message in an exception raised by Jinja isn't helpful, catch it, generate a helpful error message and re-raise with the helpful message. MkDocs will then catch that exception and log the message along with the page it is associated with.

This is what I'm doing already. The problem is that it's not only the message that is not helpful: it's the whole traceback.

Consider the following example:

- with my draft above:

```console
% make docs
poetry run mkdocs build -s
INFO    -  Cleaning site directory 
INFO    -  Building documentation to directory: /media/data/dev/pawamoy/mkdocstrings/site 
ERROR   -  Error reading page 'reference/extension.md': mkdocstrings.extension: mismatched tag: line 456, column 5 (character f):
<p></f>
If your Markdown contains angle brackets < >, try to wrap them between backticks `< >`, or replace them with &lt; and &gt; 

Aborted build due to error.
make: *** [Makefile:58: docs] Error 1
```

- with the current behavior:

```console
% make docs
poetry run mkdocs build -s
INFO    -  Cleaning site directory 
INFO    -  Building documentation to directory: /media/data/dev/pawamoy/mkdocstrings/site 
ERROR   -  Error reading page 'reference/extension.md': mkdocstrings.extension: mismatched tag: line 456, column 5 (character f):
<p></f>
If your Markdown contains angle brackets < >, try to wrap them between backticks `< >`, or replace them with &lt; and &gt; 
Traceback (most recent call last):
  File "/media/data/dev/pawamoy/mkdocstrings/src/mkdocstrings/extension.py", line 165, in run
    as_xml = XML(rendered)
  File "/home/pawamoy/.basher-packages/pyenv/pyenv/versions/3.6.9/lib/python3.6/xml/etree/ElementTree.py", line 1314, in XML
    parser.feed(text)
xml.etree.ElementTree.ParseError: mismatched tag: line 456, column 5

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/bin/mkdocs", line 11, in <module>
    load_entry_point('mkdocs==1.1', 'console_scripts', 'mkdocs')()
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/mkdocs/__main__.py", line 152, in build_command
    build.build(config.load_config(**kwargs), dirty=not clean)
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/mkdocs/commands/build.py", line 277, in build
    _populate_page(file.page, config, files, dirty)
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/mkdocs/commands/build.py", line 175, in _populate_page
    page.render(config, files)
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/mkdocs/structure/pages.py", line 183, in render
    self.content = md.convert(self.markdown)
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/markdown/core.py", line 265, in convert
    root = self.parser.parseDocument(self.lines).getroot()
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/markdown/blockparser.py", line 90, in parseDocument
    self.parseChunk(self.root, '\n'.join(lines))
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/markdown/blockparser.py", line 105, in parseChunk
    self.parseBlocks(parent, text.split('\n\n'))
  File "/home/pawamoy/.cache/pypoetry/virtualenvs/mkdocstrings-ytlBmpdO-py3.6/lib/python3.6/site-packages/markdown/blockparser.py", line 123, in parseBlocks
    if processor.run(parent, blocks) is not False:
  File "/media/data/dev/pawamoy/mkdocstrings/src/mkdocstrings/extension.py", line 181, in run
    raise ParseError(message)
xml.etree.ElementTree.ParseError: mkdocstrings.extension: mismatched tag: line 456, column 5 (character f):
<p></f>
If your Markdown contains angle brackets < >, try to wrap them between backticks `< >`, or replace them with &lt; and &gt;
make: *** [Makefile:58: docs] Error 1
```

**Comment by pawamoy** [comment]
> but not at the cost of debugging

Note that I'm not suggesting that MkDocs itself use that error. It is exactly designed for this particular use-case: a plugin developer wants to abort the build without printing the traceback, because the traceback is not useful. There's not even debugging to do: the error was handled properly.

**Comment by pawamoy** [comment]
> I would think opening background processes in a context manager (using with) would resolve that. 

I don't think I can use a context manager. The process must stay opened during the whole build process, for a performance reason.

> But I can't think of any reason to invalidate caches. For that matter, it's not clear to me why you need to open background processes either. Clearly you are attempting something much more complex that anything I ever envisioned.

My plugin can use arbitrary programs to extract structured data for source code (documentation) and inject in the generated documentation. It's an "autodoc" plugin that is language agnostic, if you want.

It's true that in my case, I don't *really need* to terminate the processes and invalidate cache, because when there is an error, MkDocs stops (even when serving in a loop), so Python will terminate all opened subprocesses, and invalidating cache is useless. But it's not really clean either. And in other cases it could be required: rolling back a change on the file system or terminating a daemonized program for example.

**Comment by pawamoy** [comment]
@waylan what about a `PluginError` instead? Something like this:

```python
class PluginError(MkDocsException):
    """Error in a plugin"""

    def __init__(self, message, reraise=True):
        super().__init__(message)
        self.reraise = reraise
```

Then catch it in the `build` function:

```python
def build(config, live_server=False, dirty=False):
    """ Perform a full site build. """
    try:
        ...
    except PluginError as error:
        # Run `build_error` plugin events.
        config['plugins'].run_event('build_error', error=e)
        if error.reraise:
            raise error
        raise SystemExit('\nAborted build due to error.')
```

This would allow plugin writers to choose between showing a traceback (`reraise=True`, the default) or not (`reraise=False`), while still running an `on_build_error` hook and properly aborting the build:

```python
# somewhere in plugin code
try:
    json.loads(output)
except JSONParseError as error:
    raise PluginError(f"{error}\n{output}", reraise=False)
```

**Comment by waylan** [comment]
A `PluginError` might make more sense. That way it would be an error unique to plugins. After all, MkDocs itself doesn't need any special exception types (except for the existing config error) and the whole purpose is to address the needs of plugins specifically. And the `reraise` attribute gives the plugin dev some control. I'm thinking the default for `reraise` should be `False` though. I realize you make prefer `True` but most plugins don't need that. At least there hasn't been a high demand for it.

I'd be happy to review a PR.
